### PR TITLE
Support mixtral long sequence 32k with bs 4

### PR DESCRIPTION
**Pull request by jychen-habana**
# What does this PR do?

Support long sequences 32k with **bs4** (move Q slicing inside loop to save memory)

**Before**
OutofMemory

**After**
**Basic Command (with --limit_hpu_graphs, --reuse_cache, --bucket_internal, --batch_size 4)**
```
QUANT_CONFIG=./quantization_config/maxabs_quant_mixtral.json python run_generation.py --model_name_or_path mistralai/Mixtral-8x7B-v0.1 --use_hpu_graphs --limit_hpu_graphs --use_kv_cache --reuse_cache --bucket_internal --bucket_size ${bucket_size} --max_new_tokens ${max_new_tokens} --bf16 --fp8 --batch_size 4 --max_input_tokens 32000
```
**Test case 1: --bucket_size 128 --max_new_tokens 128**
input 4: ('He got all',)
output 1: ('He got all the way to the top of the mountain, but he couldn’t get over the last hurdle.\n\nThe 2018 World Cup is over for Cristiano Ronaldo.\n\nThe Portugal star was unable to help his team advance to the quarterfinals, as Uruguay defeated Portugal 2-1 in the round of 16 on Saturday.\n\nRonaldo, 33, has never won a World Cup. He’s never even made it to the semifinals.\n\nRonaldo has won the Champions League five times, the Ballon d’Or',)

Stats:
Throughput (including tokenization) = **8.158104216323297 tokens/second**
Number of HPU graphs                = 337
Memory allocated                    = 69.49 GB
Max memory allocated                = 94.41 GB
Total memory available              = 94.62 GB
Graph compilation duration          = 464.618659114989 seconds

**Test case 2: --bucket_size 256 --max_new_tokens 512**
input 4: ('He got all',)
output 1: ('He got all the way to the top of the mountain, but he couldn’t get over the last hurdle.\n\nThe 2018 World Cup is over for Cristiano Ronaldo.\n\nThe Portugal star was unable to help his team advance to the quarterfinals, as Uruguay defeated Portugal 2-1 in the round of 16 on Saturday.\n\nRonaldo, 33, has never won a World Cup. He’s never even made it to the semifinals.\n\nRonaldo has won the Champions League five times, the Ballon d’Or five times, the European Championship once, and the European Golden Shoe four times.\n\nBut he’s never won a World Cup.\n\nRonaldo has scored 85 goals in 154 appearances for Portugal. He’s scored 450 goals in 438 appearances for Real Madrid.\n\nBut he’s never won a World Cup.\n\nRonaldo has scored 15 goals in 17 appearances for Portugal in World Cup qualifying. He’s scored 15 goals in 14 appearances for Portugal in the World Cup.\n\nBut he’s never won a World Cup.\n\nRonaldo has scored 15 goals in 17 appearances for Portugal in World Cup qualifying. He’s scored 15 goals in 14 appearances for Portugal in the World Cup.\n\nBut he’s never won a World Cup.\n\nRonaldo has scored 15 goals in 17 appearances for Portugal in World Cup qualifying. He’s scored 15 goals in 14 appearances for Portugal in the World Cup.\n\nBut he’s never won a World Cup.\n\nRonaldo has scored 15 goals in 17 appearances for Portugal in World Cup qualifying. He’s scored 15 goals in 14 appearances for Portugal in the World Cup.\n\nBut he’s never won a World Cup.\n\nRonaldo has scored 15 goals in 17 appearances for Portugal in World Cup qualifying. He’s scored 15 goals in 14 appearances for Portugal in the World Cup.\n\nBut he’s never won a World Cup.\n\nRonaldo has scored 15 goals in 17 appearances for Portugal in World Cup qualifying.',)

Stats:
Throughput (including tokenization) = **27.221884640574576 tokens/second**
Number of HPU graphs                = 369
Memory allocated                    = 69.68 GB
Max memory allocated                = 94.58 GB
Total memory available              = 94.62 GB
Graph compilation duration          = 547.4480734140379 seconds

**Test case 3: --bucket_size 256 --max_new_tokens 700**
input 4: ('He got all',)
output 1: ('He got all the way to the top of the mountain, but he couldn’t get over the last hurdle.\n\nThe 2018 World Cup is over for Cristiano Ronaldo.\n\nThe Portugal star was unable to help his team advance to the quarterfinals, as Uruguay defeated Portugal 2-1 in the round of 16 on Saturday.\n\nRonaldo, 33, has never won a World Cup. He’s never even made it to the semifinals.\n\nRonaldo has won the Champions League five times, the Ballon d’Or five times, the European Championship once, and the European Golden Shoe four times.\n\nBut he’s never won a World Cup.\n\nRonaldo has scored 85 goals in 154 appearances for Portugal. He’s scored 450 goals in 438 appearances for Real Madrid.\n\nBut he’s never won a World Cup.\n\nRonaldo has scored 15 goals in 17 appearances for Portugal in World Cup qualifying. He’s scored 15 goals in 14 appearances for Portugal in the World Cup.\n\nBut he’s never won a World Cup.\n\nRonaldo has scored 15 goals in 17 appearances for Portugal in World Cup qualifying. He’s scored 15 goals in 14 appearances for Portugal in the World Cup.\n\nBut he’s never won a World Cup.\n\nRonaldo has scored 15 goals in 17 appearances for Portugal in World Cup qualifying. He’s scored 15 goals in 14 appearances for Portugal in the World Cup.\n\nBut he’s never won a World Cup.\n\nRonaldo has scored 15 goals in 17 appearances for Portugal in World Cup qualifying. He’s scored 15 goals in 14 appearances for Portugal in the World Cup.\n\nBut he’s never won a World Cup.\n\nRonaldo has scored 15 goals in 17 appearances for Portugal in World Cup qualifying. He’s scored 15 goals in 14 appearances for Portugal in the World Cup.\n\nBut he’s never won a World Cup.\n\nRonaldo has scored 15 goals in 17 appearances for Portugal in World Cup qualifying. He’s scored 15 goals in 14 appearances for Portugal in the World Cup.\n\nBut he’s never won a World Cup.\n\nRonaldo has scored 15 goals in 17 appearances for Portugal in World Cup qualifying. He’s scored 15 goals in 14 appearances for Portugal in the World Cup.\n\nBut he’s never won a World Cup.\n\nRonaldo has scored 15 goals in 17 appearances for Portugal in World Cup qualifying. He’s scored 15 goals in 14 appearances for Portugal in the World Cup.\n\nBut he’s never won a World Cup.\n\nRonaldo has scored 15 goals in 17 appearances for Portugal in World Cup qualifying. He’s scored 15 goals in 14 appearances for Portugal in the World Cup.',)

Stats:
Throughput (including tokenization) = **34.39802785860282 tokens/second**
Number of HPU graphs                = 401
Memory allocated                    = 69.78 GB
Max memory allocated                = 94.6 GB
Total memory available              = 94.62 GB
Graph compilation duration          = 670.873160321964 seconds

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by jychen-habana** [comment]
Break PR https://github.com/huggingface/optimum-habana/pull/836 into small pieces, based on PR https://github.com/huggingface/optimum-habana/pull/901

**Comment by mandy-li** [comment]
@jychen-habana , please test rope_scaling with Mixtral and update the results here.

File: optimum/habana/transformers/models/mixtral/modeling_mixtral.py [hunk 1]
```diff
@@ -260,42 +309,62 @@ def forward(
...
-            import habana_frameworks.torch.hpu as ht
-
-            if q_len == 1:
-                # next token
-                with ht.sdp_kernel(enable_recompute=False):
-                    attn_output = FusedSDPA.apply(
-                        query_states, key_states, value_states, attention_mask, 0.0, False, None
-                    )
+            # support long sequences exceeding 8192
+            if not self.training and q_len == key_states.size(-2) and q_len > 8192:
```
Comment (libinta) [inline optimum/habana/transformers/models/mixtral/modeling_mixtral.py:353]: why this is hardcoded to 8192

Comment (jychen-habana) [reply optimum/habana/transformers/models/mixtral/modeling_mixtral.py:353]: For Mixtral, in BF16 case, persistent memory would be 88GB (including KV Cache with BS=1 and T=8192), and Runtime mem allocated would be like 4GB (attn part), so total mem allocated would be at least 92GB, almost close

**Comment by jychen-habana** [comment]
> @jychen-habana , please test rope_scaling with Mixtral and update the results here.

Run with rope_scaling (add below to config.json): 
`"rope_scaling": {"type":"linear","factor":2.0},`

**Test case 1: --bucket_size 1024 --max_new_tokens 512 --batch_size 1**
Input/outputs:
input 1: ('DeepSpeed is a machine learning framework',)
output 1: ('**DeepSpeed is a machine learning framework that is developed by Microsoft to train large models with billions of parameters. It is a library that is built on top of PyTorcharm and is designed to train large models with billions of parameters.** It is a library is built on top of PyTorcharm and is designed to train large models with billions of parameters. It is library is built on top Pycharm and designed to train large models with billions of parameters. It is library is built on Pycharm designed to train large models billions of parameters. It is library built on Pyarm to train large models billions parameters. It built Pyarm to train models billions. It arm train models. It train.\n\n\nDeepSpeed is a machine learning framework is developed by Microsoft to train large models with billions of parameters. It is a library is built on top PyTorarm is designed to train large models with billions parameters. It is library built on Pyarm is designed to train large models billions. It library is built arm to train billions. It is built to train.\n\n\nDeepSpeed is a machine framework is developed by Microsoft to train models billions. It is library is built Pyarm to train billions. It is built arm to train.\n\n\nDeep is machine framework developed Microsoft train billions. is library arm train.\n\n\nDeep is framework Microsoft billions.\n\n\nDeep is Microsoft\n\n\n...',)

Stats:
Throughput (including tokenization) = 18.225460417191343 tokens/second
Number of HPU graphs                = 342
Memory allocated                    = 51.91 GB
Max memory allocated                = 94.56 GB
Total memory available              = 94.62 GB
Graph compilation duration          = 314.0399896269664 seconds

**Comment by jychen-habana** [comment]
@regisss @libinta @mandy-li please help review and merge this PR, thanks!

**Review by regisss — APPROVED** [review]

**Pull request merged**